### PR TITLE
✨ Add AWS VPC Endpoints

### PR DESCRIPTION
**Pull request by mariuskimmina**
#1775

### CODE REVIEW

File: providers/aws/resources/aws_vpc.go [hunk 1]
```diff
@@ -101,6 +101,50 @@ func (a *mqlAws) getVpcs(conn *connection.AwsConnection) []*jobpool.Job {
...
+		}
+
+		for _, endpoint := range endpointsRes.VpcEndpoints {
+			var subnetIds []interface{}
+			for _, subnet := range endpoint.SubnetIds {
+				subnetIds = append(subnetIds, subnet)
+			}
+			mqlEndpoint, err := CreateResource(a.MqlRuntime, "aws.vpc.endpoint",
+				map[string]*llx.RawData{
+					"id":             llx.StringData(*endpoint.VpcEndpointId),
```
Comment (vjeffrey) [inline providers/aws/resources/aws_vpc.go]: i haven't tested this one yet, but same comment as the other pr applies - if you can please add in the arn and have that be returned by the id function. if the arn isn't returned on the response, you can assemble it (there are a bunch of examples of this in the aws_ec2 resources  file

**Comment by mariuskimmina** [comment]
I have added the arn filed, I noticed a bug tho. 
Currently I have 3 subnets in AWS, the output in cnquery looks like this

```
aws.vpcs.where[0].subnets: [
  0: aws.vpc.subnet id="subnet-021eb62bc781a5ed5"
  1: aws.vpc.subnet id="subnet-021eb62bc781a5ed5"
  2: aws.vpc.subnet id="subnet-021eb62bc781a5ed5"
]
```

So the number of subnets is correct but they all have the same id. 
I copied this part from the exisiting routeTables function and found that the same bug exists there as well

I have 2 route tables in my aws account

```
cnquery> aws.vpcs.where( arn == "arn:aws:vpc:eu-central-1:455954587013:id/vpc-091edd43468ca60e5" )[0].routeTables
aws.vpcs.where[0].routeTables: [
  0: aws.vpc.routetable id="rtb-0d28b6f55dfe767c2"
  1: aws.vpc.routetable id="rtb-0d28b6f55dfe767c2"
]
```

I see two route tables here but they both have the same id

**Comment by mariuskimmina** [comment]
Do endpoints have their own arn? I can't find any example of one online.

**Comment by mariuskimmina** [comment]
Added the id function with `Id.Data` instead since there doesn't seem to be an arn. Also I saw `Id.Data` also being used in other places across the code base such as 

```
func (a *mqlAwsS3BucketGrant) id() (string, error) {
	return a.Id.Data, nil
}
```

**Comment by vjeffrey** [comment]
this is working well!
if there's no arn for endpoints, could you prefix it with the region, so we return region/id ? 
the point of the id function here is to give the object a unique identifier. since we query across regions by default, adding region can help us ensure we don't confuse one object for another for multi-region resources

**Comment by mariuskimmina** [comment]
There seem to be a lot of resources that just return `a.Id.Data` and I couldn't find a case where we add the region to those yet. I agree that it would make sense to add the region here and also to all the other cases where we return `Id.Data` as I don't think the uniqueness of ids is guaranteed across regions (at least I couldn't find any such guarantee).

**Comment by mariuskimmina** [comment]
I have added the region. I did not test this part yet as I am now on my work notebook and don't have an AWS account setup. 
To get the region of the endpoint I just passed the region of the VPC through 

```
"region":         llx.StringData(a.Region.Data),
```

I suppose that's fine?

**Review by vjeffrey — APPROVED** [review]

File: providers/aws/resources/aws.lr [hunk 2]
```diff
@@ -62,6 +64,24 @@ private aws.vpc.routetable @defaults("id") {
   routes []dict
 }
 
+// Amazon Virtual Private Cloud (VPC) Endpoint
+private aws.vpc.endpoint @defaults("id") {
+  // Unique ID of the endpoint
+  id string
+  // type of the endpoint
```
Comment (misterpantz) [inline providers/aws/resources/aws.lr]:
```suggestion
  // Type of the endpoint
```

File: providers/aws/resources/aws.lr [hunk 3]
```diff
@@ -62,6 +64,24 @@ private aws.vpc.routetable @defaults("id") {
...
+  type string
+  // VPC the endpoint exists in
+  vpc string
+  // Region the VPC exists in
+  region string
+  // The name of the endpoint service
+  serviceName string
+  // The policy document associated with the endpoint, if applicable.
```
Comment (misterpantz) [inline providers/aws/resources/aws.lr]:
```suggestion
  // The policy document associated with the endpoint, if applicable
```

File: providers/aws/resources/aws.lr [hunk 4]
```diff
@@ -62,6 +64,24 @@ private aws.vpc.routetable @defaults("id") {
...
+  policyDocument string
+  // the subnets for the (interface) endpoint
```
Comment (misterpantz) [inline providers/aws/resources/aws.lr]:
```suggestion
  // The subnets for the (interface) endpoint
```

**Comment by tas50** [comment]
@mariuskimmina can you attach the output this resource provides so I can use it in the release notes?

**Comment by mariuskimmina** [comment]
@tas50 

```
cnquery> aws.vpcs.where( arn == "arn:aws:vpc:eu-central-1:455954587013:id/vpc-091edd43468ca60e5" )[0].endpoints.first.policyDocument
aws.vpcs.where[0].endpoints.first.policyDocument: "{
  \"Statement\": [
    {
      \"Action\": \"*\",
      \"Effect\": \"Allow\",
      \"Principal\": \"*\",
      \"Resource\": \"*\"
    }
  ]
}"
```


```
cnquery> aws.vpcs.where( arn == "arn:aws:vpc:eu-central-1:455954587013:id/vpc-091edd43468ca60e5" )[0].endpoints.first.type
aws.vpcs.where[0].endpoints.first.type: "Interface"
```


```
cnquery> aws.vpcs.where( arn == "arn:aws:vpc:eu-central-1:455954587013:id/vpc-091edd43468ca60e5" )[0].endpoints.first.serviceName
aws.vpcs.where[0].endpoints.first.serviceName: "com.amazonaws.eu-central-1.ec2"
```

Hmm, adding the region to the id didn't seem to have worked tho :thinking: 
```
aws.vpcs.where( arn == "arn:aws:vpc:eu-central-1:455954587013:id/vpc-091edd43468ca60e5" )[0].endpoints.first.id
aws.vpcs.where[0].endpoints.first.id: "vpce-0211b655551cf8b01"
```

**Comment by mariuskimmina** [comment]
Now I got the `region/id` thing to work 

```
cnquery> aws.vpcs[12].endpoints.first
! using builtin provider for aws
aws.vpcs[12].endpoints.first: aws.vpc.endpoint id="eu-central-1/vpce-05052690cb8ca48ac"
cnquery> aws.vpcs[12].endpoints.first.id
aws.vpcs[12].endpoints.first.id: "eu-central-1/vpce-05052690cb8ca48ac"
```

**Comment by mariuskimmina** [comment]
I believe this one should be ready now @vjeffrey

**Comment by vjeffrey** [comment]
👍  this is looking good, thank you! can you rebase to resolve the conflicts? after that we can merge

**Comment by mariuskimmina** [comment]
> 👍 this is looking good, thank you! can you rebase to resolve the conflicts? after that we can merge

Done

**Pull request merged**